sdks/java/io/hbase/src/main/java/org/apache/beam/sdk/io/hbase/HBaseIO.java

```diff
@@ -0,0 +1,681 @@
+import org.apache.hadoop.hbase.client.BufferedMutator;
+import org.apache.hadoop.hbase.client.BufferedMutatorParams;
+import org.apache.hadoop.hbase.client.Connection;
+import org.apache.hadoop.hbase.client.ConnectionFactory;
+import org.apache.hadoop.hbase.client.Mutation;
+import org.apache.hadoop.hbase.client.RegionLocator;
+import org.apache.hadoop.hbase.client.Result;
+import org.apache.hadoop.hbase.client.ResultScanner;
+import org.apache.hadoop.hbase.client.Scan;
+import org.apache.hadoop.hbase.client.Table;
+import org.apache.hadoop.hbase.filter.Filter;
+import org.apache.hadoop.hbase.util.Bytes;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A bounded source and sink for HBase.
+ *
+ * <p>For more information, see the online documentation at
+ * <a href="https://hbase.apache.org/">HBase</a>.
+ *
+ * <h3>Reading from HBase</h3>
+ *
+ * <p>The HBase source returns a set of rows from a single table, returning a
+ * {@code PCollection<Result>}.
+ *
+ * <p>To configure a HBase source, you must supply a table id and a {@link Configuration}
+ * to identify the HBase instance. By default, {@link HBaseIO.Read} will read all rows in the
+ * table. The row range to be read can optionally be restricted using with a {@link Scan} object
+ * or using the {@link HBaseIO.Read#withKeyRange}, and a {@link Filter} using
+ * {@link HBaseIO.Read#withFilter}, for example:
+ *
+ * <pre>{@code
+ * // Scan the entire table.
+ * p.apply("read",
+ *     HBaseIO.read()
+ *         .withConfiguration(configuration)
+ *         .withTableId("table"));
+ *
+ * // Filter data using a HBaseIO Scan
+ * Scan scan = ...
+ * p.apply("read",
+ *     HBaseIO.read()
+ *         .withConfiguration(configuration)
+ *         .withTableId("table"))
+ *         .withScan(scan));
+ *
+ * // Scan a prefix of the table.
+ * ByteKeyRange keyRange = ...;
+ * p.apply("read",
+ *     HBaseIO.read()
+ *         .withConfiguration(configuration)
+ *         .withTableId("table")
+ *         .withKeyRange(keyRange));
+ *
+ * // Scan a subset of rows that match the specified row filter.
+ * p.apply("filtered read",
+ *     HBaseIO.read()
+ *         .withConfiguration(configuration)
+ *         .withTableId("table")
+ *         .withFilter(filter));
+ * }</pre>
+ *
+ * <h3>Writing to HBase</h3>
+ *
+ * <p>The HBase sink executes a set of row mutations on a single table. It takes as input a
+ * {@link PCollection PCollection&lt;KV&lt;ByteString, Iterable&lt;Mutation&gt;&gt;&gt;}, where the
+ * {@link ByteString} is the key of the row being mutated, and each {@link Mutation} represents an
+ * idempotent transformation to that row.
+ *
+ * <p>To configure a HBase sink, you must supply a table id and a {@link Configuration}
+ * to identify the HBase instance, for example:
+ *
+ * <pre>{@code
+ * Configuration configuration = ...
+ * PCollection<KV<ByteString, Iterable<Mutation>>> data = ...;
+ * data.setCoder(HBaseIO.WRITE_CODER)
```

Comment:
Best for the end, this one I really need you to give me some hints.
Notice that to make the API similar to Bigtable, I let this Mutation there, however HBase mutation is not Serializable. So I need to encode it to pass it, but I didn't find how to do it, because if the user is doing it externally he can use a CoderRegistry or do the setCoder in the PCollection.
I don't know how to solve it otherwise because I would need something like the opposite of getDefaultOutputCoder. I mean, like getDefaultInputCoder, and I didn't find this in the API or is there any other way to do this that I am missing, can you give me some ideas ?